### ISSUE TITLE
finetune on self-defined dataset

### ISSUE THREAD

**Issue by DemonDamon** [wenet-e2e/wenet#1397]
Here is my scenario:
I want to transfer asr ability to medical field, such as symptoms descriptions recognition, which involves hundreds of thousands of medical terminologies, and I found that just adding context bias, doesn't work well. So i collect some audios to create small wenet format dataset (maybe no more than 2 hours audios), the units number is only 1215. data.list looks like below:
![Screenshot from 2022-08-22 11-43-15](https://user-images.githubusercontent.com/20499644/185834305-29d4d088-6c7b-4197-a847-658ee16cd73c.png)

Then I used wenetspeech training parameters template [https://github.com/wenet-e2e/wenet/blob/main/examples/wenetspeech/s0/conf/train_conformer_bidecoder.yaml](url), and I changed lr=0.0001 and warmup_steps=300. I also loaded wenetspeech pretrained checkpoint parameters without classification heads, onto my model. the training cli is like below:
`python train.py --gpu 1 --config ./train_conformer_bidecoder.yaml --data_type raw --symbol_table ./medical_units.txt --train_data ./train_data.list --cv_data ./cv_data.list --model_dir ./model_saved/20220822 --checkpoint ./wenetspeech-cn-conformer-checkpoint/20220506/final.pt --cmvn ./global_cmvn --num_workers 8
`
**Note: I didn't use any wenetspeech data audios in my training process**

Here is my train loss and cv loss graph:
![1-47-25](https://user-images.githubusercontent.com/20499644/185834690-7426e678-94d3-485b-b4ab-a580dd7c018e.png)

and I loaded trained final.pt to recognize test dataset, almost wrong. However, train dataset, almost correct. this is my recognition cli:
`python recognize.py --config ./model_saved/20220822/train.yaml --test_data ./test-audios/20220727/data.list --data_type raw --gpu 0 --checkpoint ./model_saved/20220822/final.pt --dict ./medical_units.txt --beam_size 20 --result_file ./recog_results_20220727.txt --batch_size 1 --mode attention`
as we can see, the test data audio recognition results almost wrong. The reason that I can come up with is **over-fitting**. And now i have no idea how to relief that. keep feeding medical audio dataset? or?
![52-10](https://user-images.githubusercontent.com/20499644/185836010-1bfa9bbf-fc78-45b9-8674-c3a5353a9d76.png)

**Comment by robin1001** [comment]
Yes, 2 hours are too few to adapt the wenetspeech model.

**Comment by DemonDamon** [comment]
> Yes, 2 hours are too few to adapt the wenetspeech model.

can i use aishell-1 pretrained model parameters instead of wenet speech for small self-defined data adaption? btw, is it correct way to transfer pretrained parameters onto medical field?

**Comment by robin1001** [comment]
aishell-1 should give poorer result than wenetspeech since the model is trained on much less data(200 vs 10,000).

You can combine the two hours data with sampled data(50~100h) from wenetspeech, and retry to finetune the model.

**Comment by DemonDamon** [comment]
> aishell-1 should give poorer result than wenetspeech since the model is trained on much less data(200 vs 10,000).
> 
> You can combine the two hours data with sampled data(50~100h) from wenetspeech, and retry to finetune the model.

how about using tts techniques to create medical audios for all medical terminologies as training dataset, instead of real-world dialogue audios, will help the final recognition above?

**Comment by DemonDamon** [comment]
> aishell-1 should give poorer result than wenetspeech since the model is trained on much less data(200 vs 10,000).
> 
> You can combine the two hours data with sampled data(50~100h) from wenetspeech, and retry to finetune the model.

when i loaded wenetspeech checkpoint to recognize test dataset, most of phoneme recognition is correct
![image](https://user-images.githubusercontent.com/20499644/186312183-3776da83-5efb-42cb-82cb-403a040915c9.png)

so for improving the issues like 
![image](https://user-images.githubusercontent.com/20499644/186313237-ac590b78-5312-4320-89bd-777c69429ced.png)

I didn't use LM provided by wenet(actually is kaldi tools), and I want to use self-defined medical audios with terminologies without any wenetspeech raw dataset, to finetune conformer-bidecoder network, by freezing some layers(shared encoders + ctc), and i couldn't get any good results
![image](https://user-images.githubusercontent.com/20499644/186319108-fd1441aa-a441-4d41-ac5e-e57c57941b61.png)
NOTE: original units.txt with 5538 chars

Q1: i consider that the whole wenet without LM is an acoustic model, is it right? 
Q2: without any wenetspeech dataset, can i finetune small medical field dataset by freezing some pretrained parameters? If Yes, which layers should I freeze? units.txt should be created on self-defined dataset or on combined units (self-defined units.txt + wenetspeech units.txt)? global_cmvn(created by self-defined datset or wenetspeech global_cmvn)?
Q3: If Q2 answer is No, can I use other LM integrated into wenet for boosting decoding results to solve issues like wrong words, wrong termonologies?

**Comment by DemonDamon** [comment]
according to basic info above, I try to freeze ctc layer + all decoder layers, and didn't add any new word into original units.txt （length of 5538）, and without any raw wenetspeech dataset, I can train with my custom dataset, the final result shows below:
![image](https://user-images.githubusercontent.com/20499644/186600030-3e43f4c4-3c1d-4afb-9eeb-8ba1409eb546.png)
it quite better than formal training, but still worse than wenetspeech pretrained model. Now i want to freeze some layers in shared encoder part, maybe some base layers, to keep the acoustic ability to common words and improve the pronunciation errors of terminology like below,
![image](https://user-images.githubusercontent.com/20499644/186602111-51d02b51-dac6-4336-b131-a10063162df4.png)

here is my question:
**Q1: ** which encoder layers should I freeze to finetune the whole network
**Q2: ** there are some units that does not contained in 5538, how to increase that without training with raw wentspeech dataset from very beginning?

@robin1001 @placebokkk

**Comment by DemonDamon** [comment]
according to basic info above, I try to freeze ctc layer + all decoder layers, and didn't add any new word into original units.txt （length of 5538）, and without any raw wenetspeech dataset, I can train with my custom dataset, the final result shows below:
![image](https://user-images.githubusercontent.com/20499644/186600030-3e43f4c4-3c1d-4afb-9eeb-8ba1409eb546.png)
it quite better than formal training, but still worse than wenetspeech pretrained model. Now i want to freeze some layers in shared encoder part, maybe some base layers, to keep the acoustic ability to common words and improve the pronunciation errors of terminology like below,
![image](https://user-images.githubusercontent.com/20499644/186602111-51d02b51-dac6-4336-b131-a10063162df4.png)

here is my question:
**Q1: ** which encoder layers should I freeze to finetune the whole network
**Q2: ** there are some units that does not contained in 5538, how to increase that without training with raw wentspeech dataset from very beginning?

@robin1001 @placebokkk

**Comment by bigcash** [comment]
my suggestion: i think you should freeze encoder, finetune ctc and decoder.

**Comment by DemonDamon** [comment]
> my suggestion: i think you should freeze encoder, finetune ctc and decoder.

yes, it works. but the result is worse than just using pretrained model directly. So I trained on overall parameters with my custom dataset without any wenetspeech corpus, but maintained original word units.txt(5538) and global_cmvn, it works so fine. But once augmenting word units length(since rarely-used word units would show up in specific field), which means some of output layer weights would not be loaded on training, the whole loss ramained big on no matter how many epochs it trained. The author suggests to add some of wenetspeech corpus on the training of augmenting units length, however I consider if exists any other solutions without using wenetspeech corpus. Any suggestions?

**Comment by bigcash** [comment]
you just need a bigger dataset...
to expand word units.txt, you can try freeze all weight, but update the weight of your new units only.

**Comment by 1215thebqtic** [comment]
> you just need a bigger dataset... to expand word units.txt, you can try freeze all weight, but update the weight of your new units only.

Hi, I use 200h custom dataset to finetune the opensource wenetspeech u2++ model, wer decreases 10% absolutely on the custom test dataset, but wer increases 10% on the wenetspeech testsets. Are there any methods to finetune the model so that it can generalise well to both custom dataset and opensource dataset? Thanks.

**Comment by Mddct** [comment]
It is almost difficult ，if the fine-tuning is performed without the participation of the original data

**Comment by bigcash** [comment]
> > you just need a bigger dataset... to expand word units.txt, you can try freeze all weight, but update the weight of your new units only.
> 
> Hi, I use 200h custom dataset to finetune the opensource wenetspeech u2++ model, wer decreases 10% absolutely on the custom test dataset, but wer increases 10% on the wenetspeech testsets. Are there any methods to finetune the model so that it can generalise well to both custom dataset and opensource dataset? Thanks.

like robin1001 say, you can combine data with sampled from wenetspeech and your owns, as new training data

**Comment by lzy473** [comment]
我想问下自己训练出来的final.pt怎么用才能够实现语音转文字呢（小白不是很会）

**Comment by lemondy** [comment]
> 我想问下自己训练出来的final.pt怎么用才能够实现语音转文字呢（小白不是很会）

可以看下 examples 目录下run.sh 脚本中如下调用 recognize.py 音频转文本的示例，里面 --checkpoint 后面用你训练的模型 pt 文件即可

https://github.com/wenet-e2e/wenet/blob/3eb9a8579d65a32b606aa04c89bdfcaca10d220b/examples/aishell/s0/run.sh#L180

**Comment by lemondy** [comment]
> > aishell-1 should give poorer result than wenetspeech since the model is trained on much less data(200 vs 10,000).
> > You can combine the two hours data with sampled data(50~100h) from wenetspeech, and retry to finetune the model.
> 
> how about using tts techniques to create medical audios for all medical terminologies as training dataset, instead of real-world dialogue audios, will help the final recognition above?

Would using TTS to generate more training data actually help improve the final recognition accuracy?

**Comment by lzy473** [comment]
我更改之后出现这样的问题
![image](https://user-images.githubusercontent.com/127758069/233769977-3a13076c-dd66-4363-afb6-b438a6347835.png)

**Comment by xingchensong** [comment]
> 我更改之后出现这样的问题 ![image](https://user-images.githubusercontent.com/127758069/233769977-3a13076c-dd66-4363-afb6-b438a6347835.png)

这个可能是torch或libtorch版本问题，可以更新一下版本

**Comment by SevenTailCat** [comment]
> > my suggestion: i think you should freeze encoder, finetune ctc and decoder.
> 
> yes, it works. but the result is worse than just using pretrained model directly. So I trained on overall parameters with my custom dataset without any wenetspeech corpus, but maintained original word units.txt(5538) and global_cmvn, it works so fine. But once augmenting word units length(since rarely-used word units would show up in specific field), which means some of output layer weights would not be loaded on training, the whole loss ramained big on no matter how many epochs it trained. The author suggests to add some of wenetspeech corpus on the training of augmenting units length, however I consider if exists any other solutions without using wenetspeech corpus. Any suggestions?

Hi, I meet the same issue. Do you have any suggestion now?
I use same of wenetspeech corpus(about 10h) on the training of augmenting units length, but after 3 epochs, there are no result after recognizing dev-set.
and the loss is like : 
![image](https://github.com/wenet-e2e/wenet/assets/16370705/5ee642de-1c7c-460c-84de-f890609e9886)
